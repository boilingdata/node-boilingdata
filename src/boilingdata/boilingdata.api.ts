/* eslint-disable @typescript-eslint/no-explicit-any */
// BoilingData API documentation
// https://www.boilingdata.com/apidoc.html

export interface IBDDataQuery {
  messageType: string; // SQL_QUERY
  requestId: string;
  sql: string;
  splitAccess?: boolean;
  splitSizeMB?: number;
  jsHooks?: {
    initFunc?: string; // (sql: string, scanCursor: number) => any;       // The return value is stored in "privCtx" and passed to other hooks as param
    headerFunc?: string; // (privCtx: any, firstRow: any) => [any, any];  // The 1st return param is the "updated privCtx"
    batchFunc?: string; // (privCtx: any, rows: any[]) => [any, any[]];   // The 1st return param is the "updated privCtx"
    footerFunc?: string; // (privCtx: any, total: number) => any;         // The 1st return param is the "updated privCtx"
    finalFunc?: string; // (privCtx: any, allRows: any[]) => any[];       // Function for transforming the whole return batch
  };
  scanCursor?: number; // offset for rows to deliver
  engine?: string; // DUCKDB (default), SQLITE
  crossRegionPolicy?: string; // DISALLOWED, ALLOWED, SELECTED
  allowedRegions?: string[];
  readCache?: string; // NONE, MEMORY_COPY, FS_COPY
  writeCache?: string; // NONE, MEMORY_COPY, FS_COPY
  outputs?: Array<{
    outputType: string; // S3, WEB_SOCKET, FILE, KAFKA
    outputPath: string;
    outputFormat: string; // PARQUET, CSV, JSON
    outputCompression: string; // NONE, SNAPPY, ZSTD, GZIP
  }>;
  tags?: Array<{
    name: string;
    value: string;
  }>;
}

export interface IBDDataResponse {
  messageType: "DATA";
  data: Array<any>;
  requestId: string;
  numOfRecords?: number;
  batchSerial?: number;
  totalBatches?: number;
  splitSerial: number;
  totalSplitSerials: number;
  subBatchSerial?: number;
  totalSubBatches?: number;
}

export enum EMessageTypes {
  SQL_QUERY = "SQL_QUERY",
}

export enum EEngineTypes {
  DUCKDB = "DUCKDB",
  SQLITE = "SQLITE",
}

export enum EEvent {
  REQUEST = "REQUEST",
  DATA = "DATA",
  INFO = "INFO",
  LOG_ERROR = "LOG_ERROR",
  LOG_WARN = "LOG_WARN",
  LOG_INFO = "LOG_INFO",
  LOG_DEBUG = "LOG_DEBUG",
  LAMBDA_EVENT = "LAMBDA_EVENT",
  QUERY_FINISHED = "QUERY_FINISHED",
  SOCKET_OPEN = "SOCKET_OPEN",
  SOCKET_CLOSED = "SOCKET_CLOSED",
}

export const globalCallbacksList = [
  "onRequest",
  "onData",
  "onInfo",
  "onLogError",
  "onLogWarn",
  "onLogInfo",
  "onLogDebug",
  "onLammbdaEvent",
  "onQueryFinished",
  "onSocketOpen",
  "onSocketClose",
];
