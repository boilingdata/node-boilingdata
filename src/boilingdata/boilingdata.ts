import { createLogger } from "bunyan";
import { BDCredentials, getBoilingDataCredentials } from "common/identity";
import { resolve } from "path";

export interface IBDCallbacks {
  onData?: () => void;
  onInfo?: () => void;
  onRequest?: () => void;
  onQueryFinished?: () => void;
  onError?: () => void;
  onLogError?: () => void;
  onlogWarn?: () => void;
  onLogInfo?: () => void;
  onLogDebug?: () => void;
  onLambdaEvent?: () => void;
  onSocketOpen?: (socketInstance: ISocketInstance) => void;
  onSocketClose?: () => void;
}

export interface IBoilingData {
  username: string;
  password: string;
  logLevel?: "trace" | "debug" | "info" | "warn" | "error" | "fatal"; // Match with Bunyan
  callbacks?: IBDCallbacks;
}

interface ISocketInstance {
  lastActivity: number;
  queries: {};
  sendQuery: (payload: any) => void;
  query: (params: any) => Promise<void>;
  bumpActivity: () => void;
  socket?: WebSocket;
  callbacks?: IBDCallbacks;
}

export class BoilingData {
  private creds?: BDCredentials;
  private socketInstance: ISocketInstance;
  private logger = createLogger({ name: "boilingdata", level: this.props.logLevel ?? "info" });

  constructor(public props: IBoilingData) {
    this.socketInstance = {
      lastActivity: Date.now(),
      callbacks: this.props.callbacks,
      sendQuery: (payload: any) => {
        this.logger.info(payload);
        // execEvent({event: "onRequest", requestId: payload.requestId, payload }, socketInstance)
        return this.socketInstance.socket?.send(JSON.stringify(payload));
      },
      bumpActivity: () => {
        this.socketInstance.lastActivity = Date.now();
      },
      queries: {},
      query: (params: any) => this.runQuery(params, this.socketInstance),
    };
  }

  public async init(): Promise<ISocketInstance> {
    return new Promise(async (resolve, _reject) => {
      const cb = this.socketInstance.callbacks;
      this.creds = await getBoilingDataCredentials(this.props.username, this.props.password);
      this.logger.debug(this.creds);
      this.socketInstance.socket = new WebSocket(this.creds.signedWebsocketUrl);
      this.socketInstance.socket.onclose = () => {
        if (!!cb?.onSocketClose) cb.onSocketClose();
      };
      this.socketInstance.socket.onopen = () => {
        if (!!cb?.onSocketOpen) cb.onSocketOpen(this.socketInstance);
        resolve(this.socketInstance);
      };
      this.socketInstance.socket.onerror = (err: any) => this.logger.error(err);
      this.socketInstance.socket.onmessage = (msg: any) => this.logger.info(msg);
    });
  }

  public async runQuery(params: any, socketInstance: ISocketInstance) {
    this.logger.info("runQuery:", params, socketInstance);
  }
}
