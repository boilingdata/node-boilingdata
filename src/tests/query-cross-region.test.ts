import { globalCallbacksList } from "../boilingdata/boilingdata.api";
import { BDAWSRegion, BoilingData } from "../boilingdata/boilingdata";

const createLogger = (_props: any): Console => console;

jest.setTimeout(60000);

const logLevel = "info";
const logger = createLogger({ name: "TEST", level: logLevel });
const username = process.env["BD_USERNAME"];
const password = process.env["BD_PASSWORD"];
if (!password || !username) throw new Error("Set BD_USERNAME and BD_PASSWORD envs");

const globalCallbacks = globalCallbacksList
  .map((cb: string) => ({ [cb]: (d: unknown) => logger.info(d) }))
  .reduce((obj, item) => ({ ...obj, ...item }), {});
globalCallbacks.onSocketOpen = () => {
  logger.info("socket open");
  return undefined;
};
globalCallbacks.onSocketClose = () => {
  logger.info("socket closed");
  return undefined;
};
const regions: BDAWSRegion[] = [
  "eu-west-1",
  "eu-north-1",
  // "eu-west-2",
  // "eu-west-3",
  // "eu-south-1",
  // "eu-central-1",
  // "us-east-1",
  "us-east-2",
  // "us-west-1",
  "us-west-2",
  // "ca-central-1",
];

describe("BD can switch the WebSocket connection endpoint dynamically", () => {
  it("run single query from 4 different supported regions", async () => {
    const expected = [
      {
        DOLocationID: 1,
        PULocationID: 1,
        RatecodeID: 5,
        VendorID: 1,
        congestion_surcharge: 0,
        extra: 0,
        fare_amount: 121,
        improvement_surcharge: 0.3,
        mta_tax: 0,
        passenger_count: 2,
        payment_type: 1,
        store_and_fwd_flag: "N",
        tip_amount: 0,
        tolls_amount: 0,
        total_amount: 121.3,
        tpep_dropoff_datetime: 1551410222000,
        tpep_pickup_datetime: 1551410199000,
        trip_distance: 0,
      },
      {
        DOLocationID: 1,
        PULocationID: 1,
        RatecodeID: 5,
        VendorID: 1,
        congestion_surcharge: 0,
        extra: 0,
        fare_amount: 110,
        improvement_surcharge: 0.3,
        mta_tax: 0,
        passenger_count: 1,
        payment_type: 1,
        store_and_fwd_flag: "N",
        tip_amount: 10,
        tolls_amount: 0,
        total_amount: 120.3,
        tpep_dropoff_datetime: 1551415659000,
        tpep_pickup_datetime: 1551415597000,
        trip_distance: 18.4,
      },
    ];

    // Run all queries in parallel
    await Promise.all(
      regions.map(async region => {
        const sql = `SELECT * FROM parquet_scan('s3://boilingdata-demo/demo2.parquet') ORDER BY DOLocationID, PULocationID, tpep_dropoff_datetime, tpep_pickup_datetime, trip_distance LIMIT 2;`;
        const bd = new BoilingData({
          username,
          password,
          globalCallbacks,
          logLevel,
          region,
        });
        await bd.connect();
        // The JSON is bigint JSON, thus converting back to "normal", to get Object properties inherited, like toString()
        const rows = JSON.parse(JSON.stringify(await bd.execQueryPromise({ sql })));
        expect(rows.sort()).toEqual(expected);
        await bd.close();
      }),
    );
  });
});

describe("BoilingData in all North-America and Europe AWS Regions", () => {
  it("runs query succesfully in other regions too", async () => {
    // Run all in parallel
    await Promise.all(
      regions.map(async region => {
        const rows: any[] = [];
        const bdInstance = new BoilingData({
          username,
          password,
          globalCallbacks,
          logLevel,
          region,
        });
        await bdInstance.connect();
        logger.info(`connected to region ${region}`);
        const bucket = region == "eu-west-1" ? "boilingdata-demo" : `${region}-boilingdata-demo`;
        const sql = `SELECT * FROM parquet_scan('s3://${bucket}/test.parquet') LIMIT 1;`;
        rows.push(await bdInstance.execQueryPromise({ sql }));
        await bdInstance.close();
        logger.info(`connection closed to region ${region}`);
        const sorted = rows.sort();
        console.log(sorted);
        expect(sorted).toMatchSnapshot();
      }),
    );
  });
});
