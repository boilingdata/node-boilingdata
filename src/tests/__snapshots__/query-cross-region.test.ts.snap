// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoilingData in all North-America and Europe AWS Regions runs query succesfully in other regions too 1`] = `
Array [
  Array [
    Object {
      "birthdate": "3/8/1971",
      "cc": "6759521864920116",
      "comments": "1E+02",
      "country": "Indonesia",
      "email": "ajordan0@com.com",
      "first_name": "Amanda",
      "gender": "Female",
      "id": 1,
      "ip_address": "1.197.201.2",
      "last_name": "Jordan",
      "registration_dttm": "2016-02-03 07:55:29+00",
      "salary": 49756.53,
      "title": "Internal Auditor",
    },
  ],
]
`;

exports[`BoilingData in all North-America and Europe AWS Regions runs query succesfully in other regions too 2`] = `
Array [
  Array [
    Object {
      "birthdate": "3/8/1971",
      "cc": "6759521864920116",
      "comments": "1E+02",
      "country": "Indonesia",
      "email": "ajordan0@com.com",
      "first_name": "Amanda",
      "gender": "Female",
      "id": 1,
      "ip_address": "1.197.201.2",
      "last_name": "Jordan",
      "registration_dttm": "2016-02-03 07:55:29+00",
      "salary": 49756.53,
      "title": "Internal Auditor",
    },
  ],
]
`;

exports[`BoilingData in all North-America and Europe AWS Regions runs query succesfully in other regions too 3`] = `
Array [
  Array [
    Object {
      "birthdate": "3/8/1971",
      "cc": "6759521864920116",
      "comments": "1E+02",
      "country": "Indonesia",
      "email": "ajordan0@com.com",
      "first_name": "Amanda",
      "gender": "Female",
      "id": 1,
      "ip_address": "1.197.201.2",
      "last_name": "Jordan",
      "registration_dttm": "2016-02-03 07:55:29+00",
      "salary": 49756.53,
      "title": "Internal Auditor",
    },
  ],
]
`;

exports[`BoilingData in all North-America and Europe AWS Regions runs query succesfully in other regions too 4`] = `
Array [
  Array [
    Object {
      "birthdate": "3/8/1971",
      "cc": "6759521864920116",
      "comments": "1E+02",
      "country": "Indonesia",
      "email": "ajordan0@com.com",
      "first_name": "Amanda",
      "gender": "Female",
      "id": 1,
      "ip_address": "1.197.201.2",
      "last_name": "Jordan",
      "registration_dttm": "2016-02-03 07:55:29+00",
      "salary": 49756.53,
      "title": "Internal Auditor",
    },
  ],
]
`;
